File: fabric/src/main/java/mcpc/tedo0627/kzeaddon/fabric/screen/SettingScreen.java
package mcpc.tedo0627.kzeaddon.fabric.screen;

import mcpc.tedo0627.kzeaddon.fabric.CustomConfig;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.gui.widget.ButtonWidget;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.text.Text;

public class SettingScreen extends Screen {

    private final CustomConfig config;

    public SettingScreen(CustomConfig config) {
        super(Text.of("message setting screen"));
        this.config = config;
    }

    @Override
    public void init() {
        MinecraftClient mc = client;
        if (mc == null) return;

        addDrawableChild(new ButtonWidget(width / 10 + 120, height / 5, 150, 20, Text.of("メッセージの設定"), (button) -> {
            mc.setScreen(new MessageSettingScreen(config, this));
        }));
        addDrawableChild(new ButtonWidget(width / 10 + 120, height / 5 + 30, 150, 20, getButtonText("リソパの再読み込みの無効化", config.isDisableResourcePackReload()), (button) -> {
            config.setDisableResourcePackReload(!config.isDisableResourcePackReload());
            button.setMessage(getButtonText("リソパの再読み込みの無効化", config.isDisableResourcePackReload()));
        }));
        addDrawableChild(new ButtonWidget(width / 10 + 120, height / 5 + 60, 150, 20, getButtonText("明るさを最大にする", mc.options.gamma >= 10000), (button) -> {
            mc.options.gamma = mc.options.gamma >= 10000 ? 1.0 : 10000;
            mc.options.write();
            button.setMessage(getButtonText("明るさを最大にする", mc.options.gamma >= 10000));
        }));
        addDrawableChild(new ButtonWidget(width / 2 - 20, height / 5 + 200, 40, 20, Text.of("閉じる"), (button) -> {
            onClose();
        }));
    }

    private Text getButtonText(String text, boolean bool) {
        return Text.of(text + ": " + (bool ? "オン" : "オフ"));
    }

    @Override
    public void render(MatrixStack matrices, int mouseX, int mouseY, float delta) {
        renderBackground(matrices);
        super.render(matrices, mouseX, mouseY, delta);
    }

    @Override
    public void onClose() {
        super.onClose();
        config.save();
    }
}

File: fabric/src/main/kotlin/mcpc/tedo0627/kzeaddon/fabric/screen/SettingScreen.kt
package mcpc.tedo0627.kzeaddon.fabric.screen

import com.mojang.blaze3d.vertex.PoseStack
import mcpc.tedo0627.kzeaddon.fabric.option.AddonOptions
import mcpc.tedo0627.kzeaddon.fabric.option.OptionConfig
import net.minecraft.client.gui.components.Button
import net.minecraft.client.gui.screens.Screen
import net.minecraft.network.chat.CommonComponents
import net.minecraft.network.chat.Component

class SettingScreen(private val previous: Screen? = null) : Screen(Component.literal("KZEAddonのオプション")) {

    override fun init() {
        mutableListOf(
            AddonOptions.hidePlayerToggle,
            AddonOptions.hidePlayerOverlay,
            AddonOptions.invisibleType,
            AddonOptions.gamma,
            AddonOptions.displayBullet,
            AddonOptions.displayKillLog,
            AddonOptions.addKillLogWeaponName,
            AddonOptions.disableKillLogWhenPressTab,
            AddonOptions.removeChatKillLog,
            AddonOptions.displayGlassTimer,
            AddonOptions.displayScoreboardTimer
        ).forEach {
            addRenderableWidget(it.createButton(minecraft?.options ?: return, getNextX(), getNextY(), 200))
        }

        addRenderableWidget(Button
            .builder(Component.translatable("kzeaddon.screen.setting.displayLocationButton")) { minecraft?.setScreen(OverlayLocationScreen(this)) }
            .bounds(getNextX(), getNextY(), 200, 20)
            .build()
        )

        addRenderableWidget(Button
            .Builder(CommonComponents.GUI_DONE) { minecraft?.setScreen(previous) }
            .pos(width / 2 - 100, height - 27)
            .size(200, 20)
            .build()
        )
    }

    private fun getNextX(): Int {
        val size = children().size
        return width / 2 + if (size % 2 == 0) - 200 - 5 else 5
    }

    private fun getNextY(): Int {
        val size = children().size
        return height / 6 - 12 + (size / 2 * 24)
    }

    override fun render(matrixStack: PoseStack, i: Int, j: Int, f: Float) {
        renderBackground(matrixStack)
        drawCenteredString(matrixStack, font, title, width / 2, 5, 0xffffff)
        super.render(matrixStack, i, j, f)
    }

    override fun onClose() {
        minecraft?.setScreen(previous)
    }

    override fun removed() {
        OptionConfig.save()
    }

    override fun isPauseScreen() = false
}